src/surface.c

```diff
@@ -0,0 +1,61 @@
+#include "surface.h"
+#include "n64sys.h"
+#include "debug.h"
+#include <assert.h>
+#include <string.h>
+
+const char* tex_format_name(tex_format_t fmt)
+{
+    switch (fmt) {
+    case FMT_NONE:   return "FMT_NONE";
+    case FMT_RGBA16: return "FMT_RGBA16";
+    case FMT_YUV16:  return "FMT_YUV16";
+    case FMT_CI4:    return "FMT_CI4";
+    case FMT_CI8:    return "FMT_CI8";
+    case FMT_IA4:    return "FMT_IA4";
+    case FMT_IA8:    return "FMT_IA8";
+    case FMT_IA16:   return "FMT_IA16";
+    case FMT_I4:     return "FMT_I4";
+    case FMT_I8:     return "FMT_I8";
+    default:         return "FMT_???";
+    }
+}
+
+surface_t surface_alloc(tex_format_t format, uint32_t width, uint32_t height)
+{
+    return (surface_t){ 
+        .flags = format | SURFACE_FLAGS_OWNEDBUFFER,
+        .width = width,
+        .height = height,
+        .stride = TEX_FORMAT_PIX2BYTES(format, width),
+        .buffer = malloc_uncached_aligned(64, height * TEX_FORMAT_PIX2BYTES(format, width)),
+    };
+}
+
+void surface_free(surface_t *surface)
+{
+    if (surface->buffer && surface->flags & SURFACE_FLAGS_OWNEDBUFFER) {
+        free_uncached(surface->buffer);
+        surface->buffer = NULL;
+    }
+    memset(surface, 0, sizeof(surface_t));
+}
+
+surface_t surface_make_sub(surface_t *parent, uint32_t x0, uint32_t y0, uint32_t width, uint32_t height)
+{
+    assert(x0 + width <= parent->width);
+    assert(y0 + height <= parent->height);
+
+    tex_format_t fmt = surface_get_format(parent);
+
+    surface_t sub;
+    sub.buffer = parent->buffer + y0 * parent->stride + TEX_FORMAT_PIX2BYTES(fmt, x0);
```

Comment:
Again, what to do if width is odd and the surface has a 4 bits per pixel format?

Reply:
It's not possible to make a sub surface starting from an odd offset in a 4bpp format, for obvious reasons. For even offsets, this should work now.

Reply:
So if `x0` is odd, then this will silently set the beginning of the buffer to the next byte. We should probably assert in that case, because this could have unforeseen consequences (like overflowing at the end of the line). And obviously the first pixel of each line is not what the user might expect.

Reply:
Done, see if it seems OK to you.

Reply:
Looks good.